src/main/java/com/cosmian/jna/findex/Findex.java

```diff
@@ -33,35 +33,59 @@ public static void upsert(
             keyPointer.write(0, key, 0, key.length);
             labelPointer.write(0, label, 0, label.length);
 
+            //
+            // Prepare outputs
+            //
+            // Allocate the amount of memory needed to store all upserted keywords.
+            byte[] newKeywordsBuffer = new byte[0];
+            IntByReference newKeywordsBufferSize = new IntByReference(newKeywordsBuffer.length);
             long start = System.currentTimeMillis();
-            // Indexes creation + insertion/update
-            unwrap(INSTANCE.h_upsert(
-                keyPointer, key.length,
-                labelPointer, label.length,
-                indexedValuesToJson(additions),
-                indexedValuesToJson(deletions),
-                entryTableNumber,
-                db.fetchEntryCallback(),
-                db.upsertEntryCallback(),
-                db.upsertChainCallback()), start);
+            int ffiCode = INSTANCE.h_upsert(newKeywordsBuffer, newKeywordsBufferSize,
+                                     keyPointer, key.length,
+                                     labelPointer, label.length,
+                                     indexedValuesToJson(additions),
+                                     indexedValuesToJson(deletions),
+                                     entryTableNumber,
+                                     db.fetchEntryCallback(),
+                                     db.upsertEntryCallback(),
+                                     db.upsertChainCallback());
+            FindexCallbackException.rethrowOnErrorCode(ffiCode, start, System.currentTimeMillis());
+
+            if (ffiCode != 0) {
+                // Retry with correct allocated size
+                newKeywordsBuffer = new byte[newKeywordsBufferSize.getValue()];
+                long startRetry = System.currentTimeMillis();
```

Comment:
the time elapsed during first call is not relevant IMHO

Reply:
I copied it from the search function, shall I remove it from there too ?

Reply:
no, I don't think so since in the case of the `search` function, real treatment is done in the first call of `h_search`